bot/utils/services.py

```diff
@@ -57,5 +113,170 @@ async def send_to_paste_service(contents: str, *, extension: str = "") -> Option
 
         log.warning(
             f"Got unexpected JSON response from paste service: {response_json}\n"
-            f"trying again ({attempt}/{FAILED_REQUEST_ATTEMPTS})."
+            f"trying again ({_attempt}/{_attempts})."
         )
+
+    return await _attempt_request(paste_url, FAILED_REQUEST_ATTEMPTS, data=contents, callback=_callback)
+
+
+async def _get_url_from_cache(url: str) -> Optional[_UnfurlReturn]:
+    """Return an unfurled URL from cache, or None if it doesn't exist or is expired."""
+    cached = await UNFURL_CACHE.get(url)
+
+    if cached is not None:
+        log.trace(f"Found hit for URL ({log}) in unfurl cache.")
+        data = json.loads(cached)
+        expiry = datetime.datetime.fromisoformat(data["expiry"])
+
+        if expiry < datetime.datetime.now(tz=datetime.timezone.utc):
+            # Cache expired, remove it and continue normally
+            log.debug(f"Cache entry for ({url}) expired, deleting.")
+            await UNFURL_CACHE.delete(url)
+        else:
+            # Found unexpired hit
+            return _UnfurlReturn(data["destination"], data["depth"], None, expiry - CACHE_LENGTH)
+
+
+async def _unfurl_url(
+    url: str,
+    redirects: int,
+    continues: int,
+    max_redirects: int,
+    max_continues: int
+) -> _JOIN_RETURNS:
+    """
+    The actual core logic of the unfurling.
+
+    Unlike the public utility, this will only follow one redirect, and parse the results.
+    The handling of the results is managed by the public function.
+
+    Returns a tuple with a final UnfurlReturn and response status,
+    or a tuple containing data to allow the caller to continue, or None.
+    """
+    # See link for documentation on how to use the worker
+    # https://github.com/python-discord/workers/tree/main/url-unfurler
+    body = {"url": url, "max-depth": max_redirects}
+    response = await _attempt_request(URLs.unfurl_worker, 3, json=body, raise_for_status=False)
+
+    if response is None:
+        return
+
+    if response.status == 200:
+        # Success, return the destination
+        content = await response.json()
+        now = datetime.datetime.now(tz=datetime.timezone.utc)
+        return _UnfurlReturn(content["destination"], redirects + content["depth"], None, now), response.status
+
+    elif response.status == 400:
+        # Unrecoverable error with the URL used for this unfurling
+        error_message = (await response.json())["error"]
+        log.warning(f"Failed to unfurl URL ({url}) with error message: {error_message}")
+        return _UnfurlReturn(error=error_message), response.status
```

Comment:
It's tempting to say that this should just raise an exception. After all, other status codes already raise exceptions thanks to `raise_for_status`, and a `_UnfurlReturn` with only the error attribute set is no more useful than an exception.